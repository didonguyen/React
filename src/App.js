import { Component } from 'react';
import AppContainer from './components/App/index.js';

class App extends Component {
  render() {
    return (
      <AppContainer />
    )
  }
}

export default App;
