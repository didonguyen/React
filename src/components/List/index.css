.list-item {
    margin: 50px 0;
}

.list__sort-group {
    padding: 8px;
}

.list__sort-btn {
    margin-right: 10px;
}

.item {
    padding: 8px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    border-bottom: solid 1px #CCC;
}

.item-title {
    font-size: 16px;
    color: #333;
    font-weight: 500;
    text-decoration: none;
}

.item-title:hover {
    color: rgb(0, 96, 128);
}

.item-author {
    font-size: 14px;
    color: #4e4e4e;
    font-weight: 400;
}

/* Pagination */
.pagination {
    padding: 10px 8px;
    background-color: #FFF;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.pagination__current {
    font-size: 16px;
    color: #333;
    font-weight: 500;
}

.interactions {
    text-align: center;
}