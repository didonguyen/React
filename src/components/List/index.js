import { Component } from 'react';
import { SubmitButton, SortButton } from '../Buttons';
import Search from '../Search';
import PropTypes from 'prop-types';

import Loading from '../Loading';

// CSS
import './index.css';

class List extends Component {
    constructor(props) {
        super(props);
        this.state = {
            searchTerm: props.searchTerm,
            filteredList: props.list,
            currentPage: props.currentPage,
            error: props.error,
            isLoading: props.isLoading,
            sortKey: props.sortKey
        }
        this.onSearchSubmit = (event) => props.onSearchSubmit(event, this.state.searchTerm);
        this.morePage = () => props.morePage(this.state.searchTerm, this.state.currentPage + 1);
        this.onSort = (sortKey) => props.onSort(sortKey);
    }

    Sort = ({ sortKey, className, activeSortKey, onSort, children }) => {
        if (sortKey === activeSortKey) {
            className=`${className} button-active`
        }
        return (
            <SortButton className={className} onClick={() => onSort(sortKey)}>
                {children}
            </SortButton>
        )
    }

    withLoading = (Component) => ({ isLoading, ...rest }) =>
        isLoading
        ? <Loading />
        : <Component {...rest} />

    ButtonWithLoading = this.withLoading(SubmitButton);

    onDismiss = (objectID) => {
        const updatedList = this.state.filteredList.filter(item => item.objectID !== objectID);
        this.setState({ filteredList: updatedList });
    }

    onClickMe = () => {
        console.log('this in onClickMe', this);
    }

    onSearchChange = (e) => {
        // const { searchTerm } = this.state;
        this.setState({ searchTerm: e.target.value });
        // const filteredList = this.state.list.filter(item => (item.title.toLowerCase().includes(searchTerm.toLowerCase())));
        // this.setState({ filteredList });
    }

    componentDidUpdate(prevProps) {
        if (prevProps !== this.props) {
            this.setState({ 
                filteredList: this.props.list,
                currentPage: this.props.currentPage,
                error: this.props.error,
                isLoading: this.props.isLoading,
                sortKey: this.props.sortKey
            })
        }
    }
    
    render() {
        const { filteredList, searchTerm, error, isLoading, sortKey } = this.state;
        const ButtonWithLoading = this.ButtonWithLoading;
        const Sort = this.Sort;
        const onSort = this.onSort;
        return (
            <div>
                <div className='search'>
                    <div className='search-title'>Search Title New</div>
                    <Search 
                        placeHolder='Search the title' 
                        className='search-input' 
                        value={searchTerm} 
                        onChange={this.onSearchChange} 
                        onSubmit={this.onSearchSubmit}
                    ></Search>
                </div>
                
                <div className='list-item'>
                    {
                        error
                        ? <div className='interactions'>
                            <p>Something went wrong.</p>
                        </div>
                        : 
                        <div>
                            <div className='list__sort-group'>
                                <Sort 
                                    sortKey='TITLE' 
                                    className='list__sort-btn'
                                    activeSortKey={sortKey}
                                    onSort={onSort}
                                >
                                    Sort By Title
                                </Sort>
                                <Sort 
                                    sortKey='AUTHOR'
                                    className='list__sort-btn'
                                    activeSortKey={sortKey}
                                    onSort={onSort}
                                >
                                    Sort By Author
                                </Sort>
                            </div>
                        {
                            (filteredList).map(item => 
                                <div key={item.objectID} className='item'>
                                    <div className='item-content'>
                                        <a href={item.url} className='item-title'>
                                            { item.title }
                                        </a>
                                        <div className='item-author'>
                                            { item.author }
                                        </div>
                                    </div>
                                    <div>
                                        <SubmitButton onClick={() => this.onDismiss(item.objectID)}>Delete</SubmitButton>
                                    </div>
                                </div>    
                            )
                        }
                        </div>
                    }
                </div>
                <div className='pagination'>
                    <div className='pagination__current'>Total Titles: {filteredList.length}</div>
                    {/* Use With Higher Order Function */}
                    <ButtonWithLoading
                        isLoading={isLoading}
                        className='pagination__button-more'
                        onClick={() => this.morePage()}
                    >
                        More Topic
                    </ButtonWithLoading>
                    {/* Use With Conditional */}
                    {/* {
                        isLoading
                        ? <Loading />
                        : <SubmitButton className='pagination__button-more' onClick={() => this.morePage()}>More Topic</SubmitButton>
                    } */}
                </div>
                {/* Click Me For Test */}
                {/* <div>
                    <SubmitButton onClick={this.onClickMe}>
                        Click me!
                    </SubmitButton>
                </div> */}
            </div>
        )
    }
}

List.propTypes = {
    list: PropTypes.arrayOf(
        PropTypes.shape({
            objectID: PropTypes.string.isRequired,
            author: PropTypes.string,
            url: PropTypes.string,
            title: PropTypes.string.isRequired,
        })
    ).isRequired,
    currentPage: PropTypes.number.isRequired,
    searchTerm: PropTypes.string.isRequired
}

export default List;