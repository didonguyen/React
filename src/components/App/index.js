import { Component } from 'react';
import axios from 'axios';
import { sortBy } from 'lodash';

import List from '../List/index.js';

import {
    DEFAULT_QUERY,
    DEFAULT_HPP,
    PATH_BASE,
    PATH_SEARCH,
    PARAM_SEARCH,
    PARAM_PAGE,
    PARAM_HPP
} from '../../constants/index.js';

// CSS
import './index.css';

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author')
}

class AppContainer extends Component {
  _isMounted = false;

  constructor(props) {
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: true,
      sortKey: 'NONE',
      isSortReverse: false
    }
  }

  onSort = (sortKey) => {
    const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
    this.setState({ sortKey, isSortReverse });
    const { searchKey, results } = this.state;
    let sortedList;
    isSortReverse 
    ? sortedList = SORTS[sortKey](this.state.results[this.state.searchKey].hits)
    : sortedList = this.state.results[this.state.searchKey].hits.reverse();
    this.setState({
      results: { 
        ...results,
        [searchKey]: {
          ...[searchKey],
          hits: sortedList
        }
      }
    })
  }

  needToSearchTopStories = (searchTerm) => {
    return !this.state.results[searchTerm];
  }

  handleMorePage = (searchTerm, nextPage) => {
    this.fetchSearchTopStories(searchTerm, nextPage);
  }

  updateSearchTopStoriesState = (hits, page) => (prevState) => {
    const { searchKey, results } = prevState;

    const oldHits = results && results[searchKey]
    ? results[searchKey].hits
    : [];

    const updatedHits = [
      ...oldHits,
      ...hits
    ];
    return {
      results: { 
        ...results,
        [searchKey]: { hits: updatedHits, page }
      },
      isLoading: false
    }
  }

  setSearchTopStories = (result) => {
    const { hits, page } = result;
    this.setState(this.updateSearchTopStoriesState(hits, page));
  }

  fetchSearchTopStories = (searchTerm, page = 0) => {
    this.setState({ isLoading: true });

    axios(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error }))
  }

  onSearchSubmit = (event, searchTermState) => {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    if (this.needToSearchTopStories(searchTermState)) {
      this.fetchSearchTopStories(searchTermState);
    }
    event.preventDefault();
  }

  componentDidMount() {
    this._isMounted = true;
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  render() {
    const { 
      searchTerm, 
      results, 
      searchKey,
      error,
      isLoading,
      sortKey
    } = this.state;

    const page = (
      results &&
      results[searchKey] &&
      results[searchKey].page
    ) || 0;

    const list = (
      results &&
      results[searchKey] &&
      results[searchKey].hits
    ) || [];

    return (
      <List
        list={list}
        searchTerm={searchTerm}
        currentPage={page}
        error={error}
        isLoading={isLoading}
        searchKey={searchKey}
        sortKey={sortKey}
        onSort={this.onSort}
        onSearchSubmit={this.onSearchSubmit}
        morePage={this.handleMorePage}
      />
    )
  }
}

export default AppContainer;
