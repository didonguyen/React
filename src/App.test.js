// import { render, screen } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from './App';
import List from './components/List';

Enzyme.configure({ adapter: new Adapter() });

describe('List', () => {
  const props = {
    list: [
      {title: '1', author: '1', objectID: '1'},
      {title: '2', author: '2', objectID: '2'},
      // {title: '3', author: '3', objectID: '3'},
    ]
  };

  it('show two items in list', () => {
    const element = shallow(
      <List {...props} />
    );

    expect(element.find('.item').length).toBe(2);
  });

})

// describe('App', () => {
//   it('reners without crashing', () => {
//     const div = document.createElement('div');
//     ReactDOM.render(<App />, div);
//     ReactDOM.unmountComponentAtNode(div);
//   });

//   test('has a valid snapshot', () => {
//     const component = renderer.create(
//       <App />
//     );
//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   })

// })


// test('renders learn react link', () => {
//   render(<App />);
//   const linkElement = screen.getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });


