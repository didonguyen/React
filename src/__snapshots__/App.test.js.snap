// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search has a valid snapshot 1`] = `
<div>
  <div
    className="search"
  >
    <div
      className="search-title"
    >
      Search Title
    </div>
    <form
      onSubmit={[Function]}
    >
      <div
        className="search-wrap"
      >
        <input
          className="search-input"
          onChange={[Function]}
          placeholder="Search the title"
          value="redux"
        />
        <button
          className="btn btn-submit"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
  <div
    className="list-item"
  />
  <div
    className="pagination"
  >
    <div
      className="pagination__current"
    >
      Total Titles: 
      0
    </div>
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      More Topic
    </button>
  </div>
</div>
`;
